### ISSUE TITLE
BUG: Unhandled error in plugin decompress: u'location'

### ISSUE THREAD

**Issue by martinberg** [Flexget/Flexget#869]
This happend with "decompress: yes". Could be a bug or could be a ID10T error. On 1.2.455 and Ubuntu 14.04

`Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flexget/task.py", line 438, in __run_plugin
    return method(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/flexget/event.py", line 22, in __call__
    return self.func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/output/decompress.py", line 275, in on_task_output
    self.handle_entry(entry, config)
  File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/output/decompress.py", line 156, in handle_entry
    archive_path = entry['location']
  File "/usr/local/lib/python2.7/dist-packages/flexget/utils/lazy_dict.py", line 69, in __getitem__
    item = self.store[key]
KeyError: u'location'``

**Comment by cvium** [comment]
It complains about the `location` field not being set, which is usually because the entry is *not* a local file. `decompress` only works on local files.

**Comment by gazpachoking** [comment]
I think @cvium has it right.

We should catch that traceback and make a nicer error though. Like 'entry does not appear to represent a local file, decompress plugin only supports entries created with filesystem plugin' or something on those lines.

**Comment by martinberg** [comment]
What do you mean by "entry"? All the files are stored locally. 
Earlier I tried with this above decompress but got the same error
`filesystem:
      path: /path/to download/folder
      recursive: yes
      regexp: '\.(rar|r0+[01]|zip)'
    decompress: yes`

But that didn't work either so I went back to this

`<task>:
    rss: <rsslink>
    series:
      720p:
        - *series*
    set:
      path: *path/to/place/files*
    decompress: yes
    template: auth`

**Comment by cvium** [comment]
Could you run your task with `--dump`? Eg. `flexget --dump execute --tasks "name of decompress task"`

**Comment by martinberg** [comment]
Running "flexget --dump execute td" gives me this. I tried without -- but It was a fail aswell.
```
usage: flexget [-h] [-V] [--test] [-c CONFIG] [--logfile LOGFILE]
               [--loglevel LEVEL] [--bugreport] [--profile [OUTFILE]] [--cron]
               [--debug-db-sessions] [--debug-warnings]
               <command> ...
flexget: error: unrecognized arguments: --dump
```

**Comment by cvium** [comment]
I guess `--dump` should be at the end then... never used it.

**Comment by martinberg** [comment]
With --dump
```
2016-02-18 20:12 VERBOSE  backlog       td              Added 4 entries from backlog
2016-02-18 20:12 VERBOSE  details       td              Produced 104 entries.
.
.
.
2016-02-18 20:12 VERBOSE  details       td              Summary - Accepted: 12 (Rejected: 0 Undecided: 92 Failed: 0)
2016-02-18 20:12 CRITICAL task          td              BUG: Unhandled error in plugin decompress: u'location'
2016-02-18 20:12 CRITICAL manager       td              An unexpected crash has occurred. Writing crash report to /home/martin/.flexget/crash_report.2016.02.18.201257681068.log. Please verify you are running the latest version of flexget by using "flexget -V" from CLI or by using version_checker plugin at http://flexget.com/wiki/Plugins/version_checker. You are currently using version 1.2.455
2016-02-18 20:12 WARNING  task          td              Aborting task (plugin: decompress)
```

**Comment by cvium** [comment]
You should probably leave out `decompress` for now :D I just want to see the files that `filesystem` finds.

**Comment by martinberg** [comment]
This is what it says before it rambles on about which ones it rejects because they have been seen before, accepts and etc. 
```
2016-02-18 20:18 VERBOSE  task_queue                    There are 1 tasks to execute. Shutdown will commence when they have completed.
2016-02-18 20:18 VERBOSE  filesystem    td              Starting to scan folders.
2016-02-18 20:18 VERBOSE  filesystem    td              Scanning folder /path/ . Recursion is set to True.
2016-02-18 20:18 VERBOSE  details       td              Produced 100 entries.
```

**Comment by cvium** [comment]
`--dump` will print all entries at the end of execution, including all fields. I'm more interested in seeing if all entries it produces contains the `location` field.

**Comment by cvium** [comment]
> But heres the full output http://pastebin.com/CFmkxEB6

None of the entries have the `location` field set. What exactly is the purpose of your task? You're using `decompress` in a task that also downloads?

**Comment by martinberg** [comment]
I may have misunderstood the whole thing then. The purpose is to decompress all he downloads of a certain task. i thought this could be done inside the download task as well. 
tl;dr want to extract files downloaded from task

**Comment by cvium** [comment]
Well, no. You're going to have to use two tasks; one that downloads and one that processes the downloaded files.

**Comment by martinberg** [comment]
Well I feel dumb now...is this something that would work? Am I thinking right?
```
templates:
  extract:
    filesystem:
      path: /media/Downloads3/Serier/Heartland/
      recursive: yes
      regexp: '\.(rar|r0+[01]|zip)'
    decompress: yes
  auth:
    *auth*
tasks:
  td:
    *download and stuff*
    template: extract
    template: auth

```

**Comment by cvium** [comment]
Almost. It should be
```
templates:
  extract:
    filesystem:
      path: /media/Downloads3/Serier/Heartland/
      recursive: yes
      regexp: '\.(rar|r0+[01]|zip)'
    decompress: yes
  auth:
    *auth*
tasks:
  td:
    *download and stuff*
    template: auth
  td2:
    template: extract
```

**Comment by martinberg** [comment]
This is what it says after I added the new task. How does the decompress work? The file ins't downloaded yet and just started downloading. Does it wait for it to complete or to next runtime for flexget?
```
2016-02-18 21:03 VERBOSE  filesystem    tdextr          Starting to scan folders.
2016-02-18 21:03 VERBOSE  filesystem    tdextr          Scanning folder /path/. Recursion is set to True.
2016-02-18 21:03 WARNING  details       tdextr          Task didn't produce any entries. This is likely due to a mis-configured or non-functional input.
2016-02-18 21:03 WARNING  task          tdextr          Task does not have any filter plugins to accept entries. You need at least one to accept the entries you  want.
2016-02-18 21:03 VERBOSE  details       tdextr          Summary - Accepted: 0 (Rejected: 0 Undecided: 0 Failed: 0)
```

**Comment by cvium** [comment]
You may need to add `accept_all: yes` to the decompress task.

**Comment by martinberg** [comment]
It is still giving me this error. 
```
Task didn't produce any entries. This is likely due to a mis-configured or non-functional input.
```
And my extract task now looks like
```
  tdextr:
    accept_all: yes
    filesystem:
      path: /media/Downloads3/Serier/Heartland/
      recursive: yes
      regexp: '\.(rar|r0+[01]|zip)'
    decompress: yes
```

**Comment by cvium** [comment]
> Task didn't produce any entries. This is likely due to a mis-configured or non-functional input.

Not necessarily an error. It could also just mean that it found nothing. You can add `no_entries_ok: yes` to suppress that message. To be sure, look for archived files in `/media/Downloads3/Serier/Heartland/`.

**Comment by martinberg** [comment]
I know I have archived files in that directory

**Comment by cvium** [comment]
Try with `regexp: '.*\.(rar|r0+[01]|zip)$'`.

**Comment by gazpachoking** [comment]
I wondered about this when decompress plugin went in. Is there a reason it shouldn't work straight in a download task? I think it'd be nice to support this, as well as fix the crashing bug in the initial report.

**Comment by martinberg** [comment]
I think I'm in love with you. It now works. Thanks for your help! 
if I want to add another path that should extract files? Do I have to declare another task then?

**Comment by cvium** [comment]
@gazpachoking Well, with the fix I've just pushed, it should do both. Although it will spam the log with warnings. Alternatively, the logging level could be debug instead.

**Comment by gazpachoking** [comment]
@cvium This kind of task works? This is what I mean about working in the download task.
```
rss: whatever
download: somewhere
decompress: yes
```

**Comment by martinberg** [comment]
@gazpachoking I can confirm it doesn't work. I tried it earlier today

**Comment by gazpachoking** [comment]
There is a possibility that it's as easy as making sure the decompress priority is after download plugin then just doing `archive_path = entry.get('location', entry.get('output'))`

**Comment by cvium** [comment]
Well, I meant it more as it would run without crashing anything, but I can see that it wouldn't really work on the stuff downloaded within the same run.

Your proposed solution should work.

**Comment by gazpachoking** [comment]
> Your proposed solution should work, although it will probably log some errors if the entries are not archives (see line 110 of the plugin).

Which makes sense to me, you tell it to decompress something that isn't compressed, it should complain. I was also wondering if there is a reason download plugin shouldn't just set 'location' field after it gets done moving the thing to the place. Might be a better solution that looking at the 'output' field.

**Comment by cvium** [comment]
`move` also sets `output`. I think both plugins should set `location` such that `location` always reflects the file's current placement.

**Comment by gazpachoking** [comment]
> move also sets output. I think both plugins should set location such that location always reflects the file's current placement.

Seems like a pretty good plan to me. I think it should still set 'output' for backwards compatibility, but I don't foresee also setting 'location' to cause any problems.

**Comment by gazpachoking** [comment]
Closing. If decompress still isn't working in a download task another enhancement ticket can be created for that.
